File: LintCode 521. Remove Duplicate Numbers in Array.js
var deduplication = function (nums) {
    var map = new Map();
    var n = nums.length;
    for ( var i = 0; i < n; i++ ) {
        if ( map.has(nums[i]) ) {
            continue;
        }
        map.set(nums[i], true);
    }

    var result = 0;
    for ( var k of map.keys() ) {
        nums[result++] = k;
    }

    return result;

};

File: LintCode 521. Remove Duplicate Numbers in Array.py
class Solution:
    """
    @param: nums: an array of integers
    @return: the number of unique integers
    """

    def deduplication(self, nums):
        lo, hi = 0, 0
        map = {}
        result = 0

        while hi < len(nums):
            if nums[hi] not in map:
                map[nums[hi]] = True
                result += 1
                nums[lo], nums[hi] = nums[hi], nums[lo]
                lo += 1
            hi += 1

        return result